python_modules/dagster/dagster/_core/storage/db_io_manager.py

```diff
@@ -207,17 +208,28 @@ def _get_table_slice(
                     )
 
                 if isinstance(context.asset_partitions_def, MultiPartitionsDefinition):
-                    multi_partition_key_mapping = cast(
-                        MultiPartitionKey, context.asset_partition_key
-                    ).keys_by_dimension
+                    partition_keys = context.asset_partition_keys
+                    dims = context.asset_partitions_def.partition_dimension_names
+                    partition_keys_by_dim: dict[str, Set[str]] = {d: set() for d in dims}
+                    # need to iterate and collect all of the partition keys by dimension so we can make the correct
+                    # time windows or lists
+                    for key in partition_keys:
+                        cast_key = cast(MultiPartitionKey, key)
+                        for d in dims:
+                            partition_keys_by_dim[d].add(cast_key.keys_by_dimension[d])
                     for part in context.asset_partitions_def.partitions_defs:
-                        partition_key = multi_partition_key_mapping[part.name]
                         if isinstance(part.partitions_def, TimeWindowPartitionsDefinition):
-                            partitions = part.partitions_def.time_window_for_partition_key(
-                                partition_key
+                            partitions = part.partitions_def.time_windows_for_partition_keys(
+                                frozenset(partition_keys_by_dim[part.name]), merge=True
                             )
+                            if len(partitions) > 1:
+                                raise DagsterInvariantViolationError(
+                                    f"Non-overlapping partition time windows {partitions} cannot be handled by the {self._io_manager_name}."
+                                )
+                            else:
+                                partitions = partitions[0]
                         else:
-                            partitions = [partition_key]
+                            partitions = list(partition_keys_by_dim[part.name])
```

Comment:
hm.... if i understand correctly, the logic here is grouping keys by dimension, and targeting all rows that match those keys, right?

I think there are situations where this would be incorrect:
I.e. for the following partitions def:
```python
MultiPartitionsDefinition(
    {
        "date": DailyPartitionsDefinition(start_date="2022-06-11"),
        "abc": StaticPartitionsDefinition(["a", "b", "c"]),
    }
)
```
the partition keys in order are:
```
['a|2022-06-11'...'a|2023-01-16', 'b|2022-06-11'...'b|2023-01-16', 'c|2022-06-11'...'c|2023-01-16']
```
which means that it's possible to execute with the range 
```python
'a|2023-01-16'...'b|2022-06-11'
```
but this range actually only contains these two partitions:
```
'a|2023-01-16'
'b|2022-06-11'
```
The existing logic would then target all rows where the 
`time dimension == 2023-01-16 or 2022-06-11` _and_ `static dimension == a or b`
which means that a row with `b|2023-01-16` would also get overwritten. Let me know if my understanding here is correct

Reply:
hmm yeah i think this could be a problem... So the way we determine the WHERE statement is that we need to get the time partitions consolidated to a single time window `[start, end]`. And then we can have a list of static partitions `[a,b,c]`. And then the WHERE clause will be 

```
WHERE time_dimension >= start AND time_dimension < end AND static_dimension in [a,b,c]
```

So the issue with the partition range you presented might be worse? It might overwrite any rows in the range `[2022-06-11, 2023-01-16]` with `a` or `b` as the static dimension. 

I think I'm maybe missing something about how multi partition ranges are constructed and how we determine if a range is valid, because I would have thought that the range `'a|2023-01-16'...'b|2022-06-11'` would have been considered invalid or something

Reply:
ah... a valid range is any contiguous subset of `multipartitions_def.get_partition_keys(...)`

Reply:
what makes a set of partitions a contiguous subset though? to me `'a|2023-01-16'...'b|2022-06-11'` looks non-contiguous... would `a` and `b` be "next to" each other be all that's required?

Reply:
if you fetch the array of partition keys via `multipartitions_def.get_partition_keys(...)`, any "block" of partition keys that are next to each other in the array is a valid range

Reply:
ah i see. i probably need to rethink the logic for how all this will work then